seahorn/trusty/user/base/lib/libc-trusty-ptr/sea_handle_table.c

```diff
@@ -0,0 +1,167 @@
+/**
+Implementation of table of handles to be used by IPC
+*/
+
+/** Global variables that keep current handle information */
+
+#include "sea_handle_table.h"
+#include "trusty_ipc.h"
+#include <malloc.h>
+#include <seahorn/seahorn.h>
+
+#define ND __declspec(noalias)
+#define MAX_HANDLES 3
+extern ND bool nd_bool(void);
+
+handle_t HANDLE_STORE[MAX_HANDLES];
+size_t g_numHandles = 0;
+
+void* malloc_can_fail(size_t sz) {
+  return nd_bool() ? malloc(sz) : NULL;
+}
+
+handle_t ptoh(HandleState* p) {
+  // This narrowing operation is safe since ptr is assumed to be aligned.
+  return p;
+}
+
+HandleState* htop(handle_t h) {
+  // This widening operation is safe since ptr is assumed to be aligned.
+  return ((HandleState*)(h));
+}
+
+
+bool sea_ht_has_msg(handle_t chan_handle) {
+ return htop(chan_handle)->msg.id > INVALID_IPC_MSG_ID;
+}
+
+uint32_t sea_ht_get_msg_id(handle_t chan_handle) {
+  return htop(chan_handle)->msg.id;
+}
+
+// assume channel handle
+void sea_ht_set_msg_id(handle_t chan_handle, uint32_t id) {
+  htop(chan_handle)->msg.id = id;
+}
+
+// assume channel handle
+size_t sea_ht_get_msg_len(handle_t chan_handle) {
+  return htop(chan_handle)->msg.len;
+}
+
+void sea_ht_set_msg_len(handle_t chan_handle, size_t len) {
+  htop(chan_handle)->msg.len = len;
+}
+
+extern ND size_t nd_msg_len(void);
+extern ND size_t nd_size(void);
+extern ND uint32_t nd_msg_id(void);
+
+void sea_ht_new_nd_msg(handle_t chan_handle) {
+ htop(chan_handle)->msg.len = nd_msg_len();
+ htop(chan_handle)->msg.id = nd_msg_id();
+ assume(htop(chan_handle)->msg.id> INVALID_IPC_MSG_ID);
+ return;
+}
+
+/**
+ * Set handle to inactive
+ */
+void sea_ht_free(handle_t handle) {
+  // we cannot "free" the handle since we might want to query state after this call
+  handle->active = false;
+}
+
+/**
+Allocates a new port handle
+
+Return INVALID_IPC_HANDLE if no handle is available to be allocated
+*/
+handle_t sea_ht_new_port(bool secure, const char *path) {
+    HandleState* h = (HandleState*) malloc_can_fail(sizeof(HandleState));
+    if (!h) return INVALID_IPC_HANDLE;
+    h->active = true;
+    h->secure = secure;
+    h->type = port;
+    h->cookie = NULL;
+    h->path = path;
+    h->msg = (struct Msg){INVALID_IPC_MSG_ID};
+    HANDLE_STORE[g_numHandles] = ptoh(h);
+    g_numHandles++;
+    return ptoh(h);
+}
+
+// TODO: should be "match"
+handle_t sea_ht_math_port(const char *path) {
+  for(size_t i = 0; i < g_numHandles; i++) {
+    if (htop(HANDLE_STORE[i])->path[0] == path[0]) return HANDLE_STORE[i];
+  }
+  return INVALID_IPC_HANDLE;
+}
+
+bool sea_contains_handle(handle_t handle) {
+  for(size_t i = 0; i < g_numHandles; i++) {
+    if (htop(HANDLE_STORE[i]) == handle) return true;
+  }
+  return false;
+}
+
+
+/**
+Non-deterministically chooses an active handle
+
+Blocks if no active handle is available
+*/
+handle_t sea_ht_choose_active_handle(void) {
+  size_t i = nd_size();
+  assume(i < g_numHandles);
+  return HANDLE_STORE[i];
```

Comment:
What if handle `HANDLE_STORE[i]` is not active? `HANDLE_STORE[i].active == false`?